### PR TITLE
ML-167: Chat moderation

### PR DESCRIPTION
**Pull request by LVala**

### CODE REVIEW

**Review by Rados13 — COMMENTED** [review]
Only removing messages from banned users left.

File: assets/js/components/event/ParticipantsList.tsx [hunk 1]
```diff
@@ -33,13 +42,24 @@ const ModeratorMenu = ({ moderatorClient, participant, eventChannel }: Moderator
...
+            value={participant.isPresenter ? "Set as a normal participant" : "Set as a presenter"}
+            className="MenuOptionText"
+          >
+            {participant.isPresenter ? "Set as a normal participant" : "Set as a presenter"}
+          </MenuItem>
+        )}
+        {!participant.isModerator && (
+          <MenuItem
+            onClick={handleClick}
+            value={participant.isBannedFromChat ? "Unban from the chat" : "Ban from the chat"}
```
[the remark this answers is not in the document]
Comment (LVala) [reply assets/js/components/event/ParticipantsList.tsx:57]: Fixed.

File: assets/js/utils/channelUtils.ts [hunk 2]
```diff
@@ -61,12 +63,24 @@ export const syncEventChannel = (
...
 
   eventChannel.push("sync_presence", {});
   return presence;
 };
 
+export const getByKey = (presence: Presence, keyEmail: string): MetasUser | undefined => {
+  let result: MetasUser | undefined;
+  presence.list((email: string, metas: Metas) => {
+    const data = metas.metas[0];
+    console.log("DATA", email, data);
```
[the remark this answers is not in the document]
Comment (LVala) [reply assets/js/utils/channelUtils.ts:78]: Fixed.

File: assets/js/components/event/ChatBox.tsx [hunk 3]
```diff
@@ -44,16 +44,20 @@ type MessageBoxProps = {
...
         let cornerClass = "";
         if (index == 0) cornerClass += "Top";
         if (index == message.messages.length - 1) cornerClass += " Bottom";
 
         return (
-          <p key={`${message.id}:${index}`} className={`SingleMessage ${cornerClass}`} lang="de">
-            {messageString}
+          <p key={messageString} className={`SingleMessage ${cornerClass}`} lang="de">
+            {index < message.moderatedNo ? <i>Moderated</i> : messageString}
           </p>
```
Comment (Rados13) [inline assets/js/components/event/ChatBox.tsx:61]: I would move this to some function.

File: assets/js/utils/useChatMessages.ts [hunk 4]
```diff
@@ -37,15 +31,43 @@ export const useChatMessages = (eventChannel: Channel | undefined): ChatMessage[
...
+        }
+
+        const newTitle = getTitle(data);
+        if (chatMessage.title != newTitle) {
+          chatMessage.title = newTitle;
+          changed = true;
+        }
+      });
+      if (changed) return [...prev];
+      else return prev;
```
Comment (Rados13) [inline assets/js/utils/useChatMessages.ts:56]:
```suggestion
      return changed ? [...prev] : prev
```

**Review by Rados13 — COMMENTED** [review]
What is the expected behavior after unbanning a user? Shouldn't messages from the user become visible again after unbanning him?

File: assets/js/components/event/ChatBox.tsx (hunk 3 above)
Comment (LVala) [reply assets/js/components/event/ChatBox.tsx]: Fixed.

File: assets/js/utils/useChatMessages.ts (hunk 4 above)
Comment (LVala) [reply assets/js/utils/useChatMessages.ts]: Fixed.

**Review by Rados13 — APPROVED** [review]

**Review by Rados13 — APPROVED** [review]

**Review by Rados13 — APPROVED** [review]

**Pull request merged**